browse-next/src/components/DeviceSetupDefineMask.vue

```diff
@@ -0,0 +1,823 @@
+    const y = points.value[i].y * canvas.value.height;
+    if (i === 0) {
+      context.moveTo(x, y);
+    } else {
+      context.lineTo(x, y);
+    }
+  }
+  context.stroke();
+
+  context.fillStyle = "rgba(13, 110, 253, 1)";
+  for (let i = 0; i < points.value.length; i++) {
+    const x = points.value[i].x * canvas.value.width;
+    const y = points.value[i].y * canvas.value.height;
+    context.beginPath();
+    context.arc(x, y, 4 * devicePixelRatio.pixelRatio.value, 0, Math.PI * 2);
+    context.closePath();
+    context.fill();
+  }
+}
+
+function isPolygonClosed(): boolean {
+  if (points.value.length < 2) {
+    return false;
+  }
+  const firstPoint = points.value[0];
+  const lastPoint = points.value[points.value.length - 1];
+  if (
+    !firstPoint ||
+    !lastPoint ||
+    !("x" in firstPoint) ||
+    !("x" in lastPoint)
+  ) {
+    return false;
+  }
+
+  const distance = Math.sqrt(
+    Math.pow(
+      firstPoint.x * canvas.value.width - lastPoint.x * canvas.value.width,
+      2
+    ) +
+      Math.pow(
+        firstPoint.y * canvas.value.height - lastPoint.y * canvas.value.height,
+        2
+      )
+  );
+
+  if (distance < polygonClosedTolerance.value) {
+    points.value[points.value.length - 1] = points.value[0];
+  }
+
+  clearMask(); //causes them to snap as desired, however recursion error
+  generateMask();
+  drawPolygon();
+  return distance < polygonClosedTolerance.value;
+}
+
+const toggleAreaSelect = () => {
+  updateExistingMaskRegions();
+  selectingArea.value = !selectingArea.value;
+  points.value = [];
+  clearMask();
+  generateMask();
+};
+
+const toggleMaskEnabled = () => {
+  maskEnabled.value = !maskEnabled.value;
+  if (maskEnabled.value) {
+    points.value = [];
+    clearMask();
+    generateMask();
+  } else {
+    clearMask();
+  }
+};
+
+const toggleInclusionRegion = () => {
+  inclusionRegion.value = !inclusionRegion.value;
+  generateMask();
+};
+
+function removePoint(): void {
+  if (points.value.length > 1) {
+    points.value.pop();
+    drawPolygon();
+  } else {
+    points.value = [];
+  }
+  clearMask();
+  generateMask();
+  drawPolygon();
+}
+
+function addRegionSelection(): void {
+  regionsArray.value.push({ regionData: points.value });
+  points.value = [];
+  toggleCreatingRegion();
+  const canvasElement = canvas.value;
+  const context = canvasElement.getContext("2d");
+  context.clearRect(0, 0, canvasElement.width, canvasElement.height);
+  generateMask();
+}
+
+function generateMask() {
+  polygonClosedTolerance.value = devicePixelRatio.pixelRatio.value * 10;
+  if (regionsArray.value.length === 0) {
+    return;
+  }
+  const maskCanvas = canvas.value;
+  const maskContext = maskCanvas.getContext("2d");
+  maskCanvas.width = cptvFrameWidth.value * devicePixelRatio.pixelRatio.value;
+  maskCanvas.height = cptvFrameHeight.value * devicePixelRatio.pixelRatio.value;
+
+  if (inclusionRegion.value) {
+    regionsArray.value.forEach((region) => {
```

Comment:
It seems like it might have been easier to keep the region data in the same format that you're saving to the backend, rather than needing a step where you convert between the two?